### ISSUE TITLE
Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by chaeunl** [IBM/aihwkit#113]
## Description
Segmentation fault (core dumped) during rebuilding a neural network model.
This is sample code:
```python
def create_model(k):
    model = AnalogSequential(AnalogLinear(784, k), nn.Sigmoid, AnalogLinear(k,10)).cuda()
return model

x = [100,200,300]
for i in x:
    ...
    model = create_model(i)
    ...
```
When I run the above code, after the first iteration, I got the message, "segmentation fault (core dumped)".
If I replace `AnalogSequential` with `nn.Sequential` and `AnalogLinear` with `nn.Linear`, it works. So, I think PyTorch and cuda are not the cause of this issue. In addition, if I run without `cuda()`, it also works.
 

## Other information

* Pytorch version: 1.7.1
* Package version: git clone
* OS: Ubuntu 18.04.2
* Python version: 3.8
* Conda version (or N/A): 
* NVIDIA Graphic Driver: 450.51.05
* CUDA Version: 11.0
* GPU: RTX 6000

**Comment by diego-plan9** [comment]
Thanks for the detailed report @chaeunl - I'm trying to replicate the issue, as we recently had a segmentation fault appearing in #112 that might have some relation.

Can you confirm the issue can be reproduced constantly (ie. is it always happening consistently at the first  iteration of the loop)?

**Comment by chaeunl** [comment]
@diego-plan9 , as I simplify my code, I didn't notice that the above code has no problem.
Instead, I found where the error comes from. I think it's about declaring the `TransferCompound` device.

This is a simplified code and I confirmed it raises error:
```python
from torch import nn
from aihwkit.nn import AnalogSequential, AnalogLinear
from aihwkit.simulator.configs import UnitCellRPUConfig
from aihwkit.simulator.configs.devices import LinearStepDevice, TransferCompound

def create_rpu(which_device="LinearStepDevice"):
    if which_device == "LinearStepDevice":
        rpu = UnitCellRPUConfig(device=LinearStepDevice())
    elif which_device == "TransferCompound":
        rpu = UnitCellRPUConfig(device=TransferCompound(unit_cell_devices=[LinearStepDevice(), LinearStepDevice()]))
    else:
        raise ValueError("Undefined Device")
    return rpu

def create_model(k, which_device="LinearStepDevice"):
    rpu = create_rpu()
    model = AnalogSequential(AnalogLinear(784, k, rpu_config=create_rpu(which_device=which_device)),nn.Sigmoid(),AnalogLinear(k, 10, rpu_config=create_rpu(which_device=which_device))).cuda()
    return model

x = [100, 200, 300]
device = "TransferCompound"
for i in x:
    model = create_model(i, which_device=device)
    print("model built")
```
If I change `device` to `"LinearStepDevice"`, it does not raise seg fault.

**Comment by diego-plan9** [comment]
Thanks @chaeunl as usual for all the detailed info: I still could not reproduce it but still have some options and configurations to try (and having the exact code and all the pointers help) - will update the issue after exhausting them.

**Comment by chaeunl** [comment]
Thank you, @diego-plan9 .
I am also trying the above code in another environment (CUDA/PyTorch are differed from this case).

**Comment by maljoras** [comment]
Hi @chaeunl,
Indeed, I now found an issue that produced a segfault for the vector device (the base class of the transfer compound). I am not completely sure whether it is the same issue that you observed, but it occurred when only one of mutliple sub devices of the vector device had diffusion or reset defined.  It is fixed in the upcoming releases.

**Comment by chaeunl** [comment]
Thank you so much, @maljoras
Is it merged into github source as of now?  As I clone the code from github directly, I can confirm it now :)

**Comment by maljoras** [comment]
@chaeunl now it is merged. You need to recompile. Let us know if you still see the segfault. Thanks!

**Comment by maljoras** [comment]
Hi @chaeunl, 
I found another issue with the CUDA `TransferCompound` in case of `transfer_every` smaller than the batch size. The update was in this case not done correctly, and in some cases (network size and batch sizes) the random states size mismatch could crash the training. Please checkout the newest version once the fix is merged.

**Comment by maljoras** [comment]
This should be fixed. Reopen if not.
